QEP 418: Better versioning for ".model3" file format (#375)
qgis/QGIS-Enhancement-Proposals: qep-418-better-versioning-model3-files.md

# QGIS Enhancement: Better versioning for ".model3" file

**Date** April 2026

**Author** Valentin Buira (@ValentinBuira)

**Contact** valentin at opengis dot ch

**Version** QGIS 4.4

## Summary

QGIS uses a “.model3” file format for storing models created using the model designer. The “.model3” despite the extension name is just an xml document under disguise.

While QGIS is evolving the “.model3” file remains static. This lead to multiple problems:

> * Deprecated or renamed parameters like  https://github.com/qgis/QGIS/commit/5bde6fec42ccc10b9fd1c4fc822fbe1a4ae05a30#diff-1a68bfeb89b1356e8ae1095b1d3576c138b2809adde3dbd2a123819c2b3a33b6R108, break the model. the idea of this QEP actually emerged after I wanted to change the type of an output from string to CRS.

Reviewer: How does the linked commit break model? From a quick look it seems the switch to a new parameter was done with backward compatibility in mind...?

Reviewer 2: I also would be interested in getting more info on how this breaks models. This approach is used in many algorithms and it was designed to keep backward compatibility.

Author: We don't expose deprecated/hidden parameters in the model designer, 

So if we take a simple example model , I made in 3.40 using the former "OPTIONS" parameter (it's just an example btw this model means nothing) 

In 3.40

<img width="991" height="496" alt="Screenshot 2026-04-14 125727" src="https://github.com/user-attachments/assets/9781c040-aac4-49b8-add3-3d3430490ed8" />

The same model in 3.44.8

<img width="975" height="451" alt="Screenshot 2026-04-14 125813" src="https://github.com/user-attachments/assets/7041c02f-d1b7-457f-a7e0-2d20d1c4d42e" />

The link to the "OPTIONS" parameter have disappear, and the actual outcome of the model is left to the details of implementation. I am not 100% sure of what happen here on execution. Is the former parameter used but not editable (if it was not connect for example and just used as a simple string )? is it ignored ? What if I save my model is the hidden parameter lost ?

In short it break the model but not the python API you can still do this in python:

```
processing.run(
    "gdal:cliprasterbyextent",
    {
        "INPUT": "C:/Users/Valentin/QField/cloud/oulipo_rennes_qfield_1/20250719_170012_modified.tif",
        "PROJWIN": "-187063.789800000,-184837.115300000,6123700.529600000,6125144.931100000 [EPSG:3857]",
        "OVERCRS": False,
        "NODATA": None,
        "OPTIONS": "COMPRESS=PACKBITS", # DEPRECATED but still work in PyQGIS running 3.44.8
        "DATA_TYPE": 0,
        "EXTRA": "",
        "OUTPUT": "TEMPORARY_OUTPUT",
    },
)
```

Reviewer: I see - so the problem is that the model looses (?) the existing connection if a parameter is hidden in the new version. In that case, my suggestion would be to have some new API within processing algorithm class that could tell the processing framework how to migrate existing connections, with several options, e.g.:
- connect to parameter X instead
- do not connect, emit a warning or an error

Author: I did more tests with manually setting a parameter of the buffer algorithm to hidden ( the tests was conducted in a private branch by just switching a parameter to hidden, no more logic was added to the processAlgorithm).

Even if not linked hidden parameters are an issue since they are executed anyway, It would be more accurate to describe them as "ghost parameters"

> new API within processing algorithm class that could tell the processing framework 
> how to migrate existing connections, with several options, 

I think it would make sense to have an API at individual level like you said, this way algorithms from plugins could also benefits from it. 

In fact _I think both this QEP and a new API would complement each others very well._ because this type of API would still break the model for older version of QGIS, we need to store info in the model we don't at the moment etc..

For the new API this raise new questions for after this QEP, should we have a internal version number per algorithms ? per provider ?

* Open a model with a feature added in later version of QGIS is broken, for example you open a model made in QGIS 3.44 that use [repeat expression](https://qgis.org/project/visual-changelogs/visualchangelog344/#feature-add-repeat-function-and-reverse-variant-for-strings) in QGIS 3.40, same could be said for any new processing algorithms

Both case silently broke the model open, without a clear warning.

> And in the first case, _we effectively silently break the promise of backward compatibility when it comes to the processing model._ 

Reviewer 3: Why is it breaking backward compatibility ? With the modification, you would still be able to load older models with new version of QGIS, and it would rely on older `OPTIONS` not the new `CREATION_OPTIONS` one.

This is because you'd like to get rid of the old `OPTIONS` one and keep only the new one ? But doing so would also break calling this algorithm using qgis_process.

Author: See my response in the other comment : https://github.com/qgis/QGIS-Enhancement-Proposals/pull/375#discussion_r3078969124

This QEP proposes to reestablish the promise of backward compatibility in the model designer. And conveniently introduce a framework for a better versioning in the “.model3” and allow us to make the format evolving as much as we want. 

> For example one thing that is missing in the “.model3” format is the dependency to plugins. If a model is using an algorithm from a plugin and the plugin is not installed. The only information we have is the name of the provider of the plugin. 

Reviewer 4: Is this addressed by this QEP?

Author: > Is this addressed by this QEP?

No but it's an example of what would be available to us. This QEP would rather focus on finding out what were the breaking changes over the last years (apart from the one from in the example), and fix them

Reviewer 4: Can you add a sentence, how this relates to this QEP (would that benefit from this framework or is it otherwise affected)?

**We would like to introduce the change as soon as possible, so we aim for 4.4 or 4.2 if possible. It’s kinda a situation of “The best time to do it was 5 years ago, the second best time is now”**

## Proposed Solution

We propose to introduce new information about the QGIS version the model was made in directly inside the “.model3” format. The idea is directly drawn from the existing logic around the .qgs file format. 

## A sense of déjà vu ? 

From the perspective of the user, the UI will be very similar of the warning already existing when opening a QGIS project (same terminology, color code, in message bar, on open). And similar to the project. Saving could effectively change the “.model3” file.

> **When loading a model created in a new version** :  “This model file was created by a newer version of QGIS (4.XX.X) and could not be completely loaded.”

Reviewer 4: ```suggestion
**When loading a model created in a new version** :  “This model file was created by a newer version of QGIS (4.XX.X) and might not be fully compatible with this version.”
```

I think we should improve this also for QGIS, this bears the meaning that we were not able to load it completely. This is not correct, it's well possible we were able to completely load it, we are just not sure.

**When loading a model created in a previous version** : “Model file is older: This model file was saved by QGIS version 4.XX.X. When saving this model file, QGIS will update it to version 4.XX.X, possibly rendering it unusable with older versions of QGIS.”

_quick'n dirty mockup to give a rough idea how that would look_ 

![](./images/qep418/mockup-user-feedback.png)

## A new key information the QGIS version

A new entry will be introduced in the file format corresponding to the version( i.e `Qgis::version()` ) in which the model was last saved. This version information will serve as the pivot point to know how to load the model file. 

> > To note the “.model3” file format already went to a format change around QGIS 3.26, as you can see with the enum `QgsProcessingModelAlgorithm::InternalVersion`. But only regarding the structure of the documents and not the version of QGIS.

Reviewer 3: > But only regarding the structure of the documents and not the version of QGIS.

Not sure to fully understand this, It seems completely related to the QGIS version when I read the enum comments

 >    Version1, //!< Created in < 3.26
  >   Version2, //!< Created in >= 3.26
 
Why not replacing InternalVersion with just the QGIS version?

Author: It's really close, as far as  I understand it, InternalVersion still maintains compatibility with both version with `switch/case` in the code to handle both version, while in this proposals a model would be transformed to latest version only

But on second thought maybe we could scrap this and rely only on the QGIS version introduced by this QEP ?

Example of versioning code in pseudocode that would be added in `QgsProcessingModelAlgorithm::loadVariant` at line ~1765

```
 
   mInternalVersion = qgsEnumKeyToValue( map.value( u"internal_version"_s ).toString(), InternalVersion::Version1 );
 
+  const QgsProjectVersion fileVersion( qgsEnumKeyToValue( map.value( u"qgis_version"_s ).toString(), "3.XX.X-Solothurn" ); );
+  const QgsProjectVersion thisVersion( Qgis::version() );
+  
+  if ( thisVersion > fileVersion )
+  {
+
+     infotext  = tr(
+                          "Model file is older: This model file was saved by QGIS version 4.XX.X.
+                          When saving this model file, QGIS will update it to version 4.XX.X, 
+                          possibly rendering it unusable with older versions of QGIS"
+    )
+                          .arg( projectVersion, Qgis::version() );
+
+    
+    push the info about the model version to the user 
+
+    if (fileVersion < version 3.44): 
+      //  Transform from old version to new version
+          fileVersion = fileVersion 4.0
+    
+    if (fileVersion <  version 4.2):

> +      //  Transform from version 4 to version 4.2

Reviewer: This is a déjà vu too 🙂 We have a mechanism for migration of QgsProject files - see QgsProjectFileTransform. Introduced around 18 years ago, there are only roughly three transforms implemented - i.e. it has ended up pretty much unused - and it has been more convenient to handle the project file backwards compatibility in the bit of the code where things are actually parsed. My gut feeling is that we would end up with the same thing here as well.

And when dealing with stuff like parameter renames, it feels like those should be handled within individual algorithm implementations anyway, rather than handled globally...

Author: > And when dealing with stuff like parameter renames, it feels like those should be handled within individual algorithm implementations anyway, rather than handled globally...

For other readers,  see in the other comment https://github.com/qgis/QGIS-Enhancement-Proposals/pull/375#discussion_r3100752820

+
+      fileVersion = thisVersion
+
+    // As much if statements as there is change in the ".model3" file format
+    
+  }
+  else if (thisVersion < fileVersion )
+  {
+    visibleMessageBar()->pushWarning( QString(), tr( "This project file was created by a newer version of QGIS (%1) and could not be completely loaded." 
).arg( projectVersion ) );
+  }
+  else
+  {
+    // versions match - nothing to do 
+  }
+  
+  // Parse normally from here on once we ensure we have a version the latest version of the file format
+
   mVariables = map.value( u"modelVariables"_s ).toMap();
   mDesignerParameterValues = map.value( u"designerParameterValues"_s ).toMap();
```

Finally QgsProcessingModelAlgorithm::toVariant will be updated to always save into the “.model3” the current version of QGIS, and the format supported by this version.

## Continuous process

> From now on, for each release where we make breaking changes to the model in QGIS we would have to add a new versioning condition. This would be ensured during review of processing related PR. 

Reviewer 4: Can you describe the "versioning condition" a bit more.

Author: Added a paragraph about change that would need to be versioned by this logic.

Since the discussion raised that the versioning of algorithms should rather happen at the level of algorithms themselves (see the discussion below). The number of format wide changes is now much lower
